### DISCUSSION TITLE
Can AssociationProxy handle updating extra columns in association objects through attribute_keyed_dict?

### DISCUSSION THREAD

**Discussion by Klikini** [sqlalchemy/sqlalchemy#11153] [Usage Questions] [answered]
I have two classes in a many-to-many relationship like this:

```python
class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[IntPK] = mapped_column(init=False)

    _consumables = relationship(
        "Consumable",
        back_populates="task",
        collection_class=attribute_keyed_dict("consumable_id"),
        cascade="all, delete-orphan",
        init=False,
    )

    consumables: AssociationProxy[Dict[str, int]] = association_proxy(
        "_consumables",
        "consumable_id",
        creator=lambda k, v: TaskConsumableUsage(consumable_id=k, qty=v),
        init=False,
    )


class Consumable(Base):
    __tablename__ = "consumables"

    id: Mapped[IntPK] = mapped_column(init=False)

    tasks = relationship("Task", back_populates="consumable", init=False)


class TaskConsumableUsage(Base):
    __tablename__ = "task_consumables"

    task_id: Mapped[int] = mapped_column(ForeignKey("tasks.id"), primary_key=True, init=False)

    consumable_id: Mapped[int] = mapped_column(ForeignKey("consumables.id"), primary_key=True)

    qty: Mapped[int] = mapped_column(default=1)

    task: Mapped[Task] = relationship(back_populates="_consumables", init=False)

    consumable: Mapped["Consumable"] = relationship(back_populates="tasks", init=False)
```

This means I'm able to do really cool things like this in my controller code:

```python
task.consumables = {"123": 5, "124": 1}  # task uses 5 of consumable #123 and 1 of consumable #124
```

That code works to create `TaskConsumableUsage` rows in the database, which are correctly populated with all 3 columns. However, if I run that code again (with the same consumables, but maybe different quantities), I get the following error:

```
duplicate key value violates unique constraint "task_consumables_pkey"
DETAIL:  Key (task_id, consumable_id)=(6, 50) already exists.
```

Is there a way to make the association proxy update existing rows by the primary key? I'm currently doing this "manually" (find and handle added/removed/updated association objects separately), so this would make it a lot easier.

Thanks for any help anyone can provide :)

Edit: using SQLAlchemy version 2.0.21 with PostgreSQL 13

**Comment by Klikini** [comment]
I was able to get this working by deleting all rows belonging to the current Task object from the association table before setting the new dict value, but it feels like kind of a hack and I still think there's a better way I don't know about.

```python
class Task(Base):
    ...

    def update_consumables(self, new_consumables: dict[int, int]) -> None:
        """
        :param new_consumables: maps consumable ID to quantity
        """
        self.consumables = {}
        session.flush()
        self.consumables = new_consumables
```

**Comment by zzzeek** [comment]
this part

```
    _consumables = relationship(
        "Consumable",
```


did you mean to say this?

```
    _consumables = relationship(
        "TaskConsumablesUsage",
```

it's supposed to delete those older TaskConsumableUsage rows, since you have delete-orphan set up.  


oh.  ~~OK, you have this problem:   #2501     so....yeah.  sorry, your workaround is exactly what we have for this right now.    if you gave TaskConsumableUsage a separate primary key column and removed the uniqueness from task_id/consumable_id, that would work, but then you are degrading your database schema.~~

edit - see next answer.  You almost have #2501 but you dont.

**Reply by zzzeek** [reply, marked as answer]
the problem you have, which I noticed right away but did not test, is that you are using strings for your integer keys.  these are integer columns, so you need to use integers:

```
task.consumables = {123: 5, 124: 1}  # task uses 5 of consumable #123 and 1 of consumable #124
task.consumables = {123: 5, 124: 2}  # task uses 5 of consumable #123 and 2 of consumable #124
```

using strings the ORM is not able to fully match up your pending TaskConsumableUsage objects to ones already loaded from the database.

when this is corrected, logic known as [row switch](https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/orm/persistence.py#L252) takes place which converts the INSERT/DELETE of those two rows into an UPDATE:

```
 UPDATE task_consumables SET consumable_id=? WHERE task_consumables.task_id = ? AND task_consumables.consumable_id = ?
2024-03-16 10:55:10,074 INFO sqlalchemy.engine.Engine [generated in 0.00018s] [(5, 1, 123), (2, 1, 124)]

```

the #2501 problem only happens when the unique key of the rows is **not** the primary key.

**Reply by zzzeek** [reply]
so that is, you have this:

```
task = Task()
task.consumables = {123: 5, 124: 1}  # task uses 5 of consumable #123 and 1 of consumable #124

s.add(task)
s.commit()

task.consumables = {123: 5, 124: 2}  # task uses 5 of consumable #123 and 2 of consumable #124

s.commit()
```

here's the SQL

```
2024-03-16 10:56:32,391 INFO sqlalchemy.engine.Engine INSERT INTO task_consumables (task_id, consumable_id, qty) VALUES (?, ?, ?)
2024-03-16 10:56:32,391 INFO sqlalchemy.engine.Engine [generated in 0.00006s] [(1, 123, 5), (1, 124, 1)]
2024-03-16 10:56:32,391 INFO sqlalchemy.engine.Engine COMMIT
2024-03-16 10:56:32,392 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-03-16 10:56:32,393 INFO sqlalchemy.engine.Engine SELECT tasks.id AS tasks_id 
FROM tasks 
WHERE tasks.id = ?
2024-03-16 10:56:32,393 INFO sqlalchemy.engine.Engine [generated in 0.00006s] (1,)
2024-03-16 10:56:32,393 INFO sqlalchemy.engine.Engine SELECT task_consumables.task_id AS task_consumables_task_id, task_consumables.consumable_id AS task_consumables_consumable_id, task_consumables.qty AS task_consumables_qty 
FROM task_consumables 
WHERE ? = task_consumables.task_id
2024-03-16 10:56:32,393 INFO sqlalchemy.engine.Engine [generated in 0.00007s] (1,)
2024-03-16 10:56:32,394 INFO sqlalchemy.engine.Engine UPDATE task_consumables SET consumable_id=? WHERE task_consumables.task_id = ? AND task_consumables.consumable_id = ?
2024-03-16 10:56:32,394 INFO sqlalchemy.engine.Engine [generated in 0.00007s] [(5, 1, 123), (2, 1, 124)]
2024-03-16 10:56:32,394 INFO sqlalchemy.engine.Engine COMMIT

```

**Reply by Klikini** [reply]
The string/int difference was absolutely the problem, but not in the way you caught:

Since posting, I had already noticed the `Dict[str, int]` and corrected it to `Dict[int, int]`, but I didn't come back and fix it here because it didn't change the behavior.

In my example code here, I set it to a dictionary explicitly, but the real application is reading it from the JSON body of an HTTP request, so I needed to say `self.consumables = {int(k): int(v) for k, v in consumables.items()}` to convert the strings to ints. Can't believe I missed that, so thank you so much for pointing out the type discrepancy, which made me double-check everything!

(And the relationship argument was a problem with my simplified code for posting here and not actually present in my code, so I apologize for the red herring.)
